services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -251,6 +258,103 @@ public URL getUrl(GetUrlRequest getUrlRequest) {
+        }
+
+        Pattern accessPointPattern = Pattern.compile("^([a-zA-Z0-9\\-]+)\\.s3-accesspoint(-fips)?(\\.dualstack)?"
+                                                     + "\\.([a-zA-Z0-9\\-]+)\\.amazonaws\\.com(.cn)?$");
+        if (accessPointPattern.matcher(uri.toString()).find()) {
+            throw SdkClientException.create("AccessPoints URI parsing is not supported");
+        }
+
+        Pattern outpostPattern = Pattern.compile("^([a-zA-Z0-9\\-]+)\\.op\\-[0-9]+\\.s3-outposts\\.([a-zA-Z0-9\\-]+)"
+                                                 + "\\.amazonaws\\.com(.cn)?$");
+        if (outpostPattern.matcher(uri.toString()).find()) {
+            throw SdkClientException.create("Outposts URI parsing is not supported");
+        }
+
+        String bucket = null;
+        String key = null;
+        String region = null;
+        boolean isPathStyle = false;
+        Map<String, List<String>> queryParams = new HashMap<>();
+        String path = uri.getPath();
+
+        if ("s3".equalsIgnoreCase(uri.getScheme())) {
+            if (uri.getAuthority() == null) {
+                throw SdkClientException.create("Invalid S3 URI: bucket not included");
+            }
+            bucket = uri.getAuthority();
+            if (path.length() > 1) {
+                key = path.substring(1);
+            }
+
+        } else {
+            if (uri.getHost() == null) {
+                throw SdkClientException.create("Invalid S3 URI: hostname not included");
+            }
+
+            Pattern endpointPattern = Pattern.compile("^(.+\\.)?s3[.-]([a-z0-9-]+)\\.");
+            Matcher matcher = endpointPattern.matcher(uri.getHost());
+            if (!matcher.find()) {
+                throw SdkClientException.create("Invalid S3 URI: hostname does not appear to be a valid S3 endpoint");
+            }
+
+            String prefix = matcher.group(1);
+            if (prefix == null || prefix.isEmpty()) {
+                isPathStyle = true;
+
+                if (!path.isEmpty() && !"/".equals(path)) {
+                    int index = path.indexOf('/', 1);
+
+                    if (index == -1) {
+                        bucket = path.substring(1);
+                    } else {
+                        bucket = path.substring(1, index);
+                        if (index != path.length() - 1) {
+                            key = path.substring(index + 1);
+                        }
+                    }
+                }
+            } else {
+                bucket = prefix.substring(0, prefix.length() - 1);
+                if (path != null && !path.isEmpty() && !"/".equals(path)) {
+                    key = path.substring(1);
+                }
+            }
+
+            if (!"amazonaws".equals(matcher.group(2))) {
+                region = matcher.group(2);
+            }
+        }
+
+        String queryPart = uri.getQuery();
+        if (queryPart != null) {
+            parseQuery(queryParams, queryPart);
```

Comment:
We can probably use `SdkHttpUtils.uriParams` to parse query params

Reply:
Hmm, doesn't `SdkHttpUtils.uriParams` decode them afterwards?

https://github.com/aws/aws-sdk-java-v2/blob/b11e9b43f2b1473248831a49786623fd13cd8a2d/utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java#L403-L408

Reply:
Yeah, I got confused since it was failing the tests that had the "=" encoded, but I removed the encoding there.

Also, uriParams() doesn't support multiple values for a query that are separated by a String, e.g., versionId=123,456,789, but only ampersand separators, .e.g, versionId=123&versionId=456&versionId=789. Do we want to support that as well? I changed to uriParams() and removed those tests for now

Reply:
I see. There's no standard behavior for multi-value query params. I think we should just support what we currently support, i.e.,`&`.